records/2023-01-26.bounce-rate-automated-emergency-brake.md

```diff
@@ -53,75 +54,78 @@ All services using the GC Notify platform are seen in aggregate by AWS. The GC N
+## Technical considerations
+### Changes to the Notification model
+In order to store the bounce rate GC Notify needs to record a few pieces of information for each notification:
+- The bounce type
+- The bounce sub-type
+- The ses_feedback_id associated with the bounce
+- The date and time the bounce occurred
 
-####  Provide bounce rate guidance
-Provide guidance to service owners to explain bounce rate concepts and how they apply to their service.  Provide guidance on how services can keep their bounce rate low. 
+This will require 4 new columns in the notifications table to store the bounce type.
 
-####  Bounce rate calculation for a service
-The formula for calculating a service's bounce rate will be the total number of hard bounces the service received divided by the total number of emails the service sent during the `[bounce-rate-window]`.  
+### Calculating the bounce rate
+We will store the bounce rate data for a service in Redis, which will be the source of truth for calculating and checking a services bounce rate.
 
-### Technical considerations
-#### Changes to the Notification model
-In order to store the bounce rate GC Notify needs to record a few pieces of information for each notification: 
-- the bounce type
-- the bounce sub-type
-- the ses_feedback_id associated with the bounce
-- the date and time the bounce occurred
-This will require 4 new columns in the notifications table to store the bounce type.  
+### Changes to the dashboard
+A new section will be added to the dashboard to present a service's bounce rate to the owner of the service. We will use the data stored in Redis to calculate and display a services bounce rate for a given `[bounce-rate-window]`.
 
-#### Calculating the bounce rate
-Since the bounce rate can nearly be calculated as-is from the data currently pulled live into the dashboard, it is possible to calculate the bounce rate on the fly in the dashboard with minimal code changes.  This avoids the need to explicitly store the bounce rate in the database or redis.
+### Monitoring bounce rates
+Bounce rates can be monitored by AWS log metrics in combination with alarms that will notify us via the `#notification-ops` channel in Slack when a service exceeds `[bounce-rate-warning-percentage]` or `[bounce-rate-critical-percentage]`.
 
-#### Changes to the dashboard
-The dashboard already pulls weekly and daily sending statistics including sending totals and total failures.  This could be extended to pull data also by `[bounce-rate-window]` to enable bounce rate calculation.  Failure data will need to be broken down further than it currently is to include the bounce type.  
+In addition to metrics and alarms, bounce rates can be monitored by a periodic celery task that polls Redis for bounce data, and notifies services that are over `[bounce-rate-warning-percentage]` and suspends services that are over `[bounce-rate-critical-percentage]`.
```

Comment:
does this need to be a celery task? we could just trigger the service suspension and notify the service owners when we receive a hard bounce that puts the service over the 10% limit. We could put that code in the same place as our log message here: https://github.com/cds-snc/notification-api/blob/main/app/celery/process_ses_receipts_tasks.py#L85

Reply:
That's a good point, and probably the direction we want to go. I'll re-write this section.

Reply:
I cleaned this section up let me know if it reflects well!

Reply:
looks good!